src/main/java/org/dataone/speedbagit/SpeedBagIt.java

```diff
@@ -119,98 +178,108 @@ public void writeToTagManifest(String path, String checksum) {
+        // Generate and add the bagit.txt file
+        InputStream bagTextStream = new ByteArrayInputStream(this.generateBagitTxt().getBytes(StandardCharsets.UTF_8));
+        this.addFile(bagTextStream, "bagit.txt", MessageDigest.getInstance(checksumAlgorithm), true);
 
-        // Now that payload-oxum is computed, add it to the bag-info.txt file
-        this.bagInfoFile = String.format("%spayload-oxum: %s\n", this.bagInfoFile, payloadOxum);
-        InputStream fileStream = new ByteArrayInputStream(this.bagInfoFile.getBytes(StandardCharsets.UTF_8));
-        this.addFile(fileStream, "bag-info.txt", MessageDigest.getInstance("MD5"), true);
+
+        // Generate and add the bag-info.txt file
+        String bagInfoFile = generateBagInfoTxt(payloadOxum, totalSize);
+        InputStream fileStream = new ByteArrayInputStream(bagInfoFile.getBytes(StandardCharsets.UTF_8));
+        this.addFile(fileStream, "bag-info.txt", MessageDigest.getInstance(checksumAlgorithm), true);
+
+        // Generate and add the data manifest file
         String fileName = String.format("manifest-%s.txt", this.checksumAlgorithm);
         fileStream = new ByteArrayInputStream(this.dataManifestFile.getBytes(StandardCharsets.UTF_8));
-        this.addFile(fileStream, fileName, MessageDigest.getInstance("MD5"), true);
+        this.addFile(fileStream, fileName, MessageDigest.getInstance(checksumAlgorithm), true);
 
         // Write all of the tag files
         for (SpeedFile streamingFile : this.tagFiles) {
-            ZipEntry entry = new ZipEntry(streamingFile.getPath());
-            zos.putNextEntry(entry);
-
-            fileStream = streamingFile.getStream();
-            fileStream.transferTo(zos);
-
+            this.streamFile(zos, streamingFile);
             String checksum = new String(streamingFile.getStream().getChecksum());
             this.writeToTagManifest(streamingFile.getPath(), checksum);
-            this.payloadOxum += streamingFile.getStream().getSize();
-            zos.closeEntry();
         }
 
-        // Write the tag-manifest
+        // Create the SpeedFile and stream it
         fileStream = new ByteArrayInputStream(this.tagManifestFile.getBytes(StandardCharsets.UTF_8));
-        //this.addFile(fileStream, "tagmanifest-md5.txt", MessageDigest.getInstance("MD5"), true);
         fileName = String.format("tagmanifest-%s.txt", this.checksumAlgorithm);
-        SpeedFile newFile = new SpeedFile(new SpeedStream(fileStream,  MessageDigest.getInstance(this.checksumAlgorithm)), fileName, true);
-
-        ZipEntry entry = new ZipEntry(newFile.getPath());
-        zos.putNextEntry(entry);
-        fileStream = newFile.getStream();
-        fileStream.transferTo(zos);
-        zos.closeEntry();
+        SpeedFile tagManifestStreamFile = new SpeedFile(new SpeedStream(fileStream,
+                MessageDigest.getInstance(this.checksumAlgorithm)), fileName, true);
+        this.streamFile(zos, tagManifestStreamFile);
```

Comment:
See my comment re `try/catch/finally` it's possible this `zos.close()` would fail to be called in the event of an exception. Close files and streams in finally blocks throughout.

Reply:
Yikes definitely an oversight on my part there. I went through and looked at all the places where the streams are opened/closed and added them to try/finally blocks to make sure that they're closed after use in ndceb38c